clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -3181,3 +3189,80 @@ LangAS CIRGenModule::getGlobalVarAddressSpace(const VarDecl *D) {
 
   return getTargetCIRGenInfo().getGlobalVarAddressSpace(*this, D);
 }
+
+mlir::StringAttr CIRGenModule::buildAnnotationString(StringRef Str) {
+  auto &Astr = AnnotationStrings[Str];
+  if (Astr)
+    return Astr;
+  Astr = builder.getStringAttr(Str);
+  return Astr;
+}
+
+mlir::ArrayAttr CIRGenModule::buildAnnotationArgs(const AnnotateAttr *Attr) {
+  ArrayRef<Expr *> Exprs = {Attr->args_begin(), Attr->args_size()};
+  if (Exprs.empty()) {
+    return mlir::ArrayAttr::get(builder.getContext(), {});
+  }
+  llvm::FoldingSetNodeID ID;
+  for (Expr *E : Exprs) {
+    ID.Add(cast<clang::ConstantExpr>(E)->getAPValueResult());
+  }
+  auto &Lookup = AnnotationArgs[ID.ComputeHash()];
+  if (Lookup)
+    return Lookup;
+
+  llvm::SmallVector<mlir::Attribute, 4> Args;
+  Args.reserve(Exprs.size());
+  for (Expr *E : Exprs) {
+    if (const auto StrE =
+            ::clang::dyn_cast<clang::StringLiteral>(E->IgnoreParenCasts())) {
+      // Add trailing null character as StringLiteral->getString() does not
+      Args.push_back(
+          buildAnnotationString(std::string(StrE->getString()) + '\0'));
```

Comment:
The extra null is done twice and we don't need any of it. You can add this LLVM requirement during LLVM lowering.